File: lemma-python/lemma_sdk/file_refs.py
"""File arguments for connector operations.

A connector operation that takes a file -- a Gmail attachment, a Drive upload
-- takes a *reference* to one, and the server reads it with the caller's own
access::

    pod.connectors.execute("gmail", "GMAIL_SEND_EMAIL", {
        "recipient_email": "ada@example.com",
        "subject": "Q3",
        "body": "Attached.",
        "attachment": pod_file("/me/reports/q3.pdf"),
    })

These only build the dict; writing it by hand is equally valid.
"""

from __future__ import annotations

from uuid import UUID


def pod_file(path: str, *, filename: str | None = None) -> dict[str, str]:
    """A file in the pod, by path -- ``/me/...`` is the caller's own folder."""
    return {"pod_path": path, **({"filename": filename} if filename else {})}


def pod_file_by_id(
    file_id: str | UUID, *, filename: str | None = None
) -> dict[str, str]:
    """A file in the pod, by id: stable across renames, and the same file for
    whoever runs the call, where ``/me`` is not."""
    return {"file_id": str(file_id), **({"filename": filename} if filename else {})}


__all__ = ["pod_file", "pod_file_by_id"]

File: lemma-typescript/src/file-refs.ts
/**
 * File arguments for connector operations.
 *
 * An operation that takes a file — a Gmail attachment, a Drive upload — takes a
 * reference to one, and the server reads it with the caller's own access:
 *
 *   client.connectors.operations.execute(scope, "GMAIL_SEND_EMAIL", {
 *     recipient_email: "anukul@lemma.work",
 *     attachment: podFile("/me/reports/q3.pdf"),
 *   });
 *
 * These only build the object; writing it by hand is equally valid.
 */
export type PodFileRef =
  | { pod_path: string; filename?: string }
  | { file_id: string; filename?: string };

/** A file in the pod, by path — `/me/...` is the caller's own folder. */
export function podFile(path: string, filename?: string): PodFileRef {
  return filename ? { pod_path: path, filename } : { pod_path: path };
}

/** A file in the pod, by id: stable across renames, and the same file for
 *  whoever runs the call, where `/me` is not. */
export function podFileById(fileId: string, filename?: string): PodFileRef {
  return filename ? { file_id: fileId, filename } : { file_id: fileId };
}
